src/shared/tpm2-util.c

```diff
@@ -7885,23 +7885,30 @@ Tpm2Support tpm2_support(void) {
                 if (r < 0) {
                         if (r != -ENOENT)
                                 log_debug_errno(r, "Unable to test whether /sys/class/tpmrm/ exists and is populated, assuming it is not: %m");
-                } else if (r == 0) /* populated! */
-                        support |= TPM2_SUPPORT_SUBSYSTEM|TPM2_SUPPORT_DRIVER;
-                else
-                        /* If the directory exists but is empty, we know the subsystem is enabled but no
-                         * driver has been loaded yet. */
-                        support |= TPM2_SUPPORT_SUBSYSTEM;
+                } else {
+                        /* If the directory exists but is empty (r > 0), we know the subsystem is enabled but
+                         * no driver has been loaded yet. */
+
+                        if (FLAGS_SET(mask, TPM2_SUPPORT_SUBSYSTEM))
+                                support |= TPM2_SUPPORT_SUBSYSTEM;
+
+                        if (FLAGS_SET(mask, TPM2_SUPPORT_DRIVER) && r == 0) /* populated! */
+                                support |= TPM2_SUPPORT_DRIVER;
+                }
         }
 
-        if (efi_has_tpm2())
+        if (FLAGS_SET(mask, TPM2_SUPPORT_FIRMWARE) && efi_has_tpm2())
                 support |= TPM2_SUPPORT_FIRMWARE;
 
 #if HAVE_TPM2
-        support |= TPM2_SUPPORT_SYSTEM;
+        if (FLAGS_SET(mask, TPM2_SUPPORT_SYSTEM))
+                support |= TPM2_SUPPORT_SYSTEM;
 
-        r = dlopen_tpm2();
-        if (r >= 0)
-                support |= TPM2_SUPPORT_LIBRARIES;
+        if (FLAGS_SET(mask, TPM2_SUPPORT_LIBRARIES)) {
+                r = dlopen_tpm2();
+                if (r >= 0)
+                        support |= TPM2_SUPPORT_LIBRARIES;
+        }
 #endif
 
         return support;
```

Comment:
or alternatively, just don't bother with individually doing `& mask` above, but instead just all all bits into `support`, but then do `return support & mask;` down here to get rid of all the bits we don't want to report at once.